### PR TITLE
Refactor: fault engine

### PR DESCRIPTION
**Pull request by Kristoff-starling**

### CODE REVIEW

**Comment by Kristoff-starling** [comment]
cc @livingshade

File: compiler/codegen/generator.py [hunk 1]
```diff
@@ -166,9 +171,81 @@ def visitSelectStatement(self, node: SelectStatement, ctx: Context):
...
+            {
+                var_name: RustVariable(
+                    var_name, node.value.accept(self.type_generator, None), False
+                )
+            }
+        )
+
+    def visitFunctionValue(self, node: FunctionValue, ctx: Context):
+        if node.value == "random":
+            func_str = "rand::random::<f32>()"
```
Comment (livingshade) [inline compiler/codegen/generator.py:204]: We should infer the type is `f32` from type of variable in the future, when we have IR.

File: compiler/codegen/generator.py [hunk 2]
```diff
@@ -166,9 +171,81 @@ def visitSelectStatement(self, node: SelectStatement, ctx: Context):
...
+        ctx.push_code(op_str)
+
+    def visitCompareOp(self, node: CompareOp, ctx: Context):
+        if node == CompareOp.LT:
+            op_str = "<"
+        else:
+            print("node", node)
+            print(node.value, CompareOp.LT.value)
+            print(node.value == CompareOp.LT.value)
+            print(node == CompareOp.LT)
```
Comment (livingshade) [inline compiler/codegen/generator.py:232]: Maybe we should delete the print and leave a error message.

File: compiler/codegen/generator.py [hunk 3]
```diff
@@ -166,9 +171,81 @@ def visitSelectStatement(self, node: SelectStatement, ctx: Context):
...
+            raise NotImplementedError
+
+        ctx.push_code(op_str)
+
+    def visitSearchCondition(self, node: SearchCondition, ctx: Context):
+        node.lvalue.accept(self, ctx)
+        lvalue_str = ctx.pop_code()
+        print(lvalue_str)
```
Comment (livingshade) [inline compiler/codegen/generator.py:240]: Maybe we should delete the print here

**Review by livingshade — CHANGES_REQUESTED** [review]
Looks good to me and I really appreciate the generalization in `select statement`! We can merge once the debug codes are removed.

File: compiler/codegen/generator.py (hunk 1 above)
Comment (Kristoff-starling) [reply compiler/codegen/generator.py:204]: Yes, we can add type inference in the future.

**Review by livingshade — APPROVED** [review]

**Pull request merged**